### PR TITLE
Tracks the anchors and stores them in the outlink metadata

### PR DESCRIPTION
**Pull request by jnioche**

### CODE REVIEW

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java [hunk 1]
```diff
@@ -61,6 +62,11 @@
 
     private MetadataTransfer metadataTransfer;
 
+    private boolean trackAnchors = true;
+
+    /** Metadata key name for tracking the anchors */
+    public static final String anchorsKeyName = "anchors";
```
Comment (GuiForget) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: Code style: ANCHORS_KEY_NAME

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java [hunk 2]
```diff
@@ -145,11 +154,15 @@ public void execute(Tuple tuple) {
...
-            slinks = new HashSet<String>(links.size());
+            slinks = new HashMap<String, List<String>>(links.size());
             for (Element link : links) {
                 String targetURL = link.attr("abs:href");
+                String anchor = link.text();
                 if (StringUtils.isNotBlank(targetURL)) {
-                    slinks.add(targetURL);
+                    List<String> anchors = slinks.get(targetURL);
+                    if (anchors == null) anchors = new LinkedList<String>();
+                    if (StringUtils.isNotBlank(anchor)) anchors.add(anchor);
```
Comment (GuiForget) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: Code style: should be within a block

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java [hunk 3]
```diff
@@ -205,14 +218,22 @@ public void execute(Tuple tuple) {
...
         }
 
-        for (String outlink : linksKept) {
+        for (String outlink : linksKept.keySet()) {
             // configure which metadata gets inherited from parent
             Metadata linkMetadata = metadataTransfer.getMetaForOutlink(url, metadata);
+            // add the anchors to the metadata?
+            if (trackAnchors) {
+                List<String> anchors = linksKept.get(outlink);
+                if (anchors != null && anchors.size() > 0) {
```
Comment (GuiForget) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: anchors can never be null

**Comment by jnioche** [comment]
Fixed and squashed.

**Comment by GuiForget** [comment]
Thanks

**Pull request merged**